src/eve-server/market/MarketProxyService.cpp

```diff
@@ -383,56 +433,86 @@ PyResult MarketProxyService::PlaceCharOrder(PyCallArgs &call, PyInt* stationID,
+                stDataMgr.GetStationName(call.client->GetStationID()).c_str()
+            );
+
+            call.client->SendErrorMsg(
+                "You cannot sell %s from %s while you are in %s.  Locations mismatch",
+                iRef->name(),
+                stDataMgr.GetStationName(stationID->value()).c_str(),
+                stDataMgr.GetStationName(call.client->GetStationID()).c_str()
+            );
+
             return nullptr;
         }
 
-        //TODO: verify orderRange against their skills.   client may do this...verify
+        //TODO: verify orderRange against their skills. client may do this...verify
 
         // they are allowed to sell this thing...
 
-        //is this standing order or immediate?
+        // is this standing order or immediate?
         if (duration->value() == 0) {
-            // immediate - loop to search and fill buy orders at or above asking price until qty depleted or no orders found
-            bool search(true);
+            bool completedOrder(false);
+
             uint32 orderID(0), origQty(quantity->value());
-            while (quantity->value() and search) {
+
+            // set an upper bound (this used to be a while loop that spun
+            // forever in some cases)
+            for (int i = 0; i < 1000; i++) {
+                _log(MARKET__DUMP, "Mkt::PlaceCharOrder(): issue 16 - finding buy order: %i, %i, %i, %.2f", typeID->value(), stationID->value(), quantity->value(), price->value());
+
                 orderID = MarketDB::FindBuyOrder(typeID->value(), stationID->value(), quantity->value(), price->value());
-                if (orderID) {
-                    _log(MARKET__TRACE, "PlaceCharOrder - Found buy order #%u in %s for %s.", \
-                            orderID, stDataMgr.GetStationName(stationID->value()).c_str(), call.client->GetName());
-                    search = sMktMgr.ExecuteBuyOrder(call.client, orderID, iRef, quantity->value(), useCorp->value(), typeID->value(), stationID->value(), price->value());
+
+                if (!orderID) {
+                    continue;
+                }
+
+                _log(MARKET__TRACE,
+                    "PlaceCharOrder - Found buy order #%u in %s for %s.",
+                    orderID,
+                    stDataMgr.GetStationName(stationID->value()).c_str(),
+                    call.client->GetName()
+                );
+
+                completedOrder = sMktMgr.ExecuteBuyOrder(
+                    call.client,
+                    orderID,
+                    iRef,
+                    quantity->value(),
+                    useCorp->value(),
+                    typeID->value(),
+                    stationID->value(),
+                    price->value()
+                );
+
+                if (!completedOrder) {
+                    continue;
                 }
+
+                _log(MARKET__DUMP, "Mkt::PlaceCharOrder(): issue 16 - order completed");
+                break;
```

Comment:
The code changes in this PR fix infinite loop issues that occurred in the previous version of this function, right here.